### PR TITLE
Introduce RandomAccess.SetLength

### LINKED ISSUE #58376: [API Proposal]: Introduce RandomAccess.SetLength(SafeFileHandle handle) method
### Background and motivation

Recently, we have introduced new `RandomAccess` type (#24847) which exposes some convenient methods for File IO. One of them is `RandomAccess.GetLength(SafeFileHandle)` that allows for reading file length.

When it comes to setting file length, users can:

* use `FileMode.Truncate` to truncate given file when opening it.
* specify `preallocationSize` to set initial size when opening the file.

The problem is that by using the new APIs, they can't set file length after the file has been opened.

`FileStream` exposes `SetLength` method, so should `RandomAccess`.

### API Proposal

```C#
namespace System.IO
{
    public static class RandomAccess
    {
        public static void SetLength(SafeFileHandle handle, long length);
    }
}
```     

### API Usage

```C#
void WriteToFile(string path, string contents, Encoding encoding)
{
    long estimatedFileSize = encoding.GetMaxByteCount(contents.Length); // estimated file length
    using SafeFileHandle fileHandle = OpenHandle(path, FileMode.Create, FileAccess.Write, preallocationSize: estimatedFileSize);

    byte[] bytes = encoding.GetBytes(contents);
    RandomAccess.Write(fileHandle, bytes, fileOffset: 0);

    if (bytes.Length < estimatedFileSize) // shrink the size if initial estimation was wrong
    {
        RandomAccess.SetLength(fileHandle, bytes.Length);
    }
}
```     

### Risks

I can't see any.

### PR DESCRIPTION
**Pull request by adamsitnik**
fixes #58376

### CODE REVIEW

File: src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+        [MemberData(nameof(GetSyncAsyncOptions))]
+        public void AllowsForShrinking(FileOptions options)
+        {
+            using (SafeFileHandle handle = File.OpenHandle(GetTestFilePath(), FileMode.CreateNew, FileAccess.Write, options: options))
+            {
+                RandomAccess.SetLength(handle, FileSize);
+                Assert.Equal(FileSize, RandomAccess.GetLength(handle));
+
+                RandomAccess.SetLength(handle, FileSize / 2);
+                Assert.Equal(FileSize / 2, RandomAccess.GetLength(handle));
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs:44]: Worth validating SetLength can set it to 0?

File: src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+            using (SafeFileHandle handle = File.OpenHandle(GetTestFilePath(), FileMode.CreateNew, FileAccess.Write, options: options))
+            {
+                RandomAccess.SetLength(handle, FileSize);
+
+                byte[] buffer = new byte[FileSize + 1];
+                Assert.Equal(FileSize, RandomAccess.Read(handle, buffer, 0));
+                Assert.All(buffer, @byte => Assert.Equal(0, @byte));
+            }
+        }
+    }
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs:61]: What about tests for error conditions? e.g. argument validation?

**Review by stephentoub — APPROVED** [review]
Other than my comments, LGTM.

File: src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs (hunk 2 above)
Comment (adamsitnik) [reply src/libraries/System.IO.FileSystem/tests/RandomAccess/SetLength.cs:64]: > What about tests for error conditions? e.g. argument validation?

This is covered by the base class: https://github.com/dotnet/runtime/blob/main/src/libraries/System.IO.FileSystem/tests/RandomAccess/Base.cs#L28

**Pull request merged**